### PR TITLE
YJDH-167: Kesäseteli Wizard Step 2

### PR DESCRIPTION
**Pull request by nenonja**
## Description :sparkles:
Kesäseteli step 2 implemented: https://app.abstract.com/projects/a9aaf5d4-fd7c-4b52-b634-7799553d9b67/branches/master/collections/2691daef-c503-44f1-b689-7014911dc50a
excluding
- attachment file inputs
- no tests written (maybe later) :( 

## Issues :bug:
Minor ui glitches like not scrolling up when clicking next button on wizard

### CODE REVIEW

File: frontend/benefit/applicant/src/components/applications/forms/application/deMinimisAid/list/DeMinimisAidsList.tsx [hunk 1]
```diff
@@ -45,9 +45,7 @@ const DeMinimisAidsList: React.FC = () => {
               grant[DE_MINIMIS_AID_KEYS.AMOUNT]
             } €`}</$GridCell>
             <$GridCell $colSpan={2}>
-              {formatDate(
-                parseDate(grant[DE_MINIMIS_AID_KEYS.GRANTED_AT], 'yyyy-MM-dd')
-              )}
+              {formatDate(parseDate(grant[DE_MINIMIS_AID_KEYS.GRANTED_AT]))}
```
Comment (nenonja) [inline frontend/benefit/applicant/src/components/applications/forms/application/deMinimisAid/list/DeMinimisAidsList.tsx:48]: parseDate now recognizes the format with regex, so no need to give format anymore.

File: frontend/kesaseteli/employer/src/components/application/form/DateInput.tsx [hunk 2]
```diff
@@ -0,0 +1,116 @@
+import { DateInput as HdsDateInput } from 'hds-react';
```
Comment (nenonja) [inline frontend/kesaseteli/employer/src/components/application/form/DateInput.tsx:1]: Represents date input component 
Screenshot 2021-09-20 at 3 43 51 PM

File: frontend/kesaseteli/employer/src/components/application/form/SelectionGroup.tsx [hunk 3]
```diff
@@ -0,0 +1,87 @@
+import { RadioButton, SelectionGroup as HdsSelectionGroup, SelectionGroupProps } from 'hds-react';
```
Comment (nenonja) [inline frontend/kesaseteli/employer/src/components/application/form/SelectionGroup.tsx:1]: represents radio button group 
Screenshot 2021-09-20 at 3 46 13 PM

File: frontend/kesaseteli/employer/src/components/application/form/TextInput.tsx [hunk 4]
```diff
@@ -1,47 +1,92 @@
-import { TextInput as HdsTextInput, TextInputProps } from 'hds-react';
```
Comment (nenonja) [inline frontend/kesaseteli/employer/src/components/application/form/TextInput.tsx:1]: Represents any text input (text, number, decimal number, textarea)

File: frontend/kesaseteli/employer/src/components/application/steps/step2/Step2Employments.tsx [hunk 5]
```diff
@@ -0,0 +1,17 @@
+import ApplicationForm from 'kesaseteli/employer/components/application/ApplicationForm';
```
Comment (nenonja) [inline frontend/kesaseteli/employer/src/components/application/steps/step2/Step2Employments.tsx:1]: top level component on step 2

File: frontend/kesaseteli/employer/src/components/application/steps/step2/accordions/AccordionActionButtons.tsx [hunk 6]
```diff
@@ -0,0 +1,74 @@
+import { IconTrash } from 'hds-react';
```
Comment (nenonja) [inline frontend/kesaseteli/employer/src/components/application/steps/step2/accordions/AccordionActionButtons.tsx:1]: Represents action buttons inside one accordion
Screenshot 2021-09-20 at 3 41 40 PM

File: frontend/kesaseteli/employer/src/components/application/steps/step2/accordions/AddNewEmploymentButton.tsx [hunk 7]
```diff
@@ -0,0 +1,31 @@
+import {
```
Comment (nenonja) [inline frontend/kesaseteli/employer/src/components/application/steps/step2/accordions/AddNewEmploymentButton.tsx:1]: represents add new employment button, which adds new accordion form.

File: frontend/kesaseteli/employer/src/components/application/steps/step2/accordions/EmploymentAccordion.tsx [hunk 8]
```diff
@@ -0,0 +1,171 @@
+import DateInput from 'kesaseteli/employer/components/application/form/DateInput';
```
Comment (nenonja) [inline frontend/kesaseteli/employer/src/components/application/steps/step2/accordions/EmploymentAccordion.tsx:1]: represents accordion form for single employment

File: frontend/kesaseteli/employer/src/components/application/steps/step2/accordions/EmploymentAccordionHeader.tsx [hunk 9]
```diff
@@ -0,0 +1,38 @@
+import { IconAlertCircleFill } from 'hds-react';
```
Comment (nenonja) [inline frontend/kesaseteli/employer/src/components/application/steps/step2/accordions/EmploymentAccordionHeader.tsx:1]: represents accordion header:
Screenshot 2021-09-20 at 3 53 21 PM
It indicates error when accordion is closed and form inside is invalid

File: frontend/kesaseteli/employer/src/components/application/steps/step2/accordions/EmploymentAccordions.tsx [hunk 10]
```diff
@@ -0,0 +1,37 @@
+import AddNewEmploymentButton from 'kesaseteli/employer/components/application/steps/step2/accordions/AddNewEmploymentButton';
```
Comment (nenonja) [inline frontend/kesaseteli/employer/src/components/application/steps/step2/accordions/EmploymentAccordions.tsx:1]: Represents the list of accordions

File: frontend/kesaseteli/employer/src/hooks/application/useAccordionStateLocalStorage.ts [hunk 11]
```diff
@@ -0,0 +1,28 @@
+import { useFormContext } from 'react-hook-form';
```
Comment (nenonja) [inline frontend/kesaseteli/employer/src/hooks/application/useAccordionStateLocalStorage.ts:1]: Remembers which accordion was closed and which one is open
Screenshot 2021-09-20 at 4 02 36 PM

File: frontend/kesaseteli/employer/src/utils/localstorage.utils.ts [hunk 12]
```diff
@@ -0,0 +1,11 @@
+// eslint-disable-next-line eslint-comments/disable-enable-pair
+/* eslint-disable scanjs-rules/identifier_localStorage */
+export const clearLocalStorage = (keyPrefix?: string): void => {
```
Comment (nenonja) [inline frontend/kesaseteli/employer/src/utils/localstorage.utils.ts:3]: localstorage is cleared when logging out or when sending application. Currently only states of accordions are saved into it.

File: frontend/shared/src/components/accordion/Accordion.tsx [hunk 13]
```diff
@@ -0,0 +1,65 @@
+import { IconAngleDown, IconAngleUp } from 'hds-react';
```
Comment (nenonja) [inline frontend/shared/src/components/accordion/Accordion.tsx:1]: This is 80% copy-paste of HDS accordion component. It did not support saving the state of being open or closed, also header needed to be customized.

File: frontend/shared/src/hooks/useAccordion.ts [hunk 14]
```diff
@@ -0,0 +1,73 @@
+import noop from 'lodash/noop';
```
Comment (nenonja) [inline frontend/shared/src/hooks/useAccordion.ts:1]: 80% copy-paste from useAccordion of the hds-component. See previous comment.

File: frontend/kesaseteli/employer/src/__tests__/utils/components/render-page.tsx [hunk 15]
```diff
@@ -7,20 +7,21 @@ import { NextRouter } from 'next/router';
 import React from 'react';
 import { QueryClient, QueryClientProvider } from 'react-query';
 import getDefaultReactQueryTestClient from 'shared/__tests__/utils/react-query/get-default-react-query-test-client';
-import { render, RenderResult } from 'shared/__tests__/utils/test-utils';
+import { act,render } from 'shared/__tests__/utils/test-utils';
```
Comment (Nizar-Rahme) [inline frontend/kesaseteli/employer/src/__tests__/utils/components/render-page.tsx:10]: 😃 
```suggestion
import { act, render } from 'shared/__tests__/utils/test-utils';
```

File: frontend/kesaseteli/employer/src/components/application/form/ActionButtons.sc.ts [hunk 16]
```diff
@@ -1,12 +1,7 @@
...
-  justify-content: space-between;
-  margin-top: ${(props) => props.theme.spacing.l};
-`;
-
-export const $ApplicationAction = styled.div`
-  display: flex;
-  flex-direction: column;
+export const $ButtonSection = styled(FormSection)`
+  margin-top: ${(props) => props.theme.spacing.m};
+  padding-bottom: ${(props) => props.theme.spacing.m};
```
Comment (Nizar-Rahme) [inline frontend/kesaseteli/employer/src/components/application/form/ActionButtons.sc.ts:6]: IMHO, we may use the `css` prop for specifying margin values instead of creating a new styled component.

```TSX
 const theme = useTheme();

  <FormSection
    css={`
      margin-top: ${theme.spacing.m};
    `}>
     ...
```

Comment (Nizar-Rahme) [reply frontend/kesaseteli/employer/src/components/application/form/ActionButtons.sc.ts:6]: afaik, `styled-components` doesn't treat these as inline styles, but it will create a dedicated component under the hood.

File: frontend/kesaseteli/employer/src/components/application/steps/step2/error-notification/EmploymentsErrorNotification.tsx [hunk 17]
```diff
@@ -0,0 +1,61 @@
+import { $Notification,$NotificationDescription } from 'kesaseteli/employer/components/application/steps/step2/error-notification/EmploymentsErrorNotification.sc';
```
Comment (Nizar-Rahme) [inline frontend/kesaseteli/employer/src/components/application/steps/step2/error-notification/EmploymentsErrorNotification.tsx:1]: 😉 
```suggestion
import { $Notification, $NotificationDescription } from 'kesaseteli/employer/components/application/steps/step2/error-notification/EmploymentsErrorNotification.sc';
```

File: frontend/kesaseteli/employer/src/components/application/steps/step2/error-notification/EmploymentsErrorNotification.sc.ts [hunk 18]
```diff
@@ -0,0 +1,9 @@
+import { Notification } from 'hds-react';
+import { $GridCell } from 'shared/components/forms/section/FormSection.sc';
+import styled from 'styled-components';
+
+export const $NotificationDescription = styled($GridCell)``;
```
Comment (Nizar-Rahme) [inline frontend/kesaseteli/employer/src/components/application/steps/step2/error-notification/EmploymentsErrorNotification.sc.ts:5]: Why is this needed?

File: frontend/shared/src/constants.ts [hunk 19]
```diff
@@ -10,3 +10,11 @@ export const POSTAL_CODE_REGEX = /^\d{5}$/;
...
   /^[\w',.ÄÅÖäåö-][^\d!#$%&()*+/:;<=>?@[\\\]_{|}~¡¿÷ˆ]+$/;
 export const CITY_REGEX = /^[ A-Za-zÄÅÖäåö-]+$/;
+
+export const EMAIL_REGEX =
+  // eslint-disable-next-line security/detect-unsafe-regex
+  /^(([^\s"(),.:;<>@[\\\]]+(\.[^\s"(),.:;<>@[\\\]]+)*)|(".+"))@((\[(?:\d{1,3}\.){3}\d{1,3}])|(([\dA-Za-z-]+\.)+[A-Za-z]{2,}))$/;
+
+// eslint-disable-next-line security/detect-unsafe-regex
+export const DATE_UI_REGEX = /^(?:\d{1,2}.){2}\d{4}$/;
+export const DATE_BACKEND_REGEX = /^\d{4}-\d{2}-\d{2}$/;
```
Comment (Nizar-Rahme) [inline frontend/shared/src/constants.ts:20]: I used in my PR more thorough versions for validating dates inspired by [this answer on Stack Over Flow](https://stackoverflow.com/a/22061879/9441362) so I didn't have to disable the security rule.
https://github.com/City-of-Helsinki/yjdh/blob/71f2fb70942dc842c7fcbaeaf198ec505cf36559/frontend/shared/src/utils/date.utils.ts#L40
https://github.com/City-of-Helsinki/yjdh/blob/71f2fb70942dc842c7fcbaeaf198ec505cf36559/frontend/shared/src/utils/date.utils.ts#L44

File: frontend/shared/src/constants.ts [hunk 20]
```diff
@@ -10,3 +10,11 @@ export const POSTAL_CODE_REGEX = /^\d{5}$/;
 export const NAMES_REGEX =
   /^[\w',.ÄÅÖäåö-][^\d!#$%&()*+/:;<=>?@[\\\]_{|}~¡¿÷ˆ]+$/;
 export const CITY_REGEX = /^[ A-Za-zÄÅÖäåö-]+$/;
+
+export const EMAIL_REGEX =
+  // eslint-disable-next-line security/detect-unsafe-regex
+  /^(([^\s"(),.:;<>@[\\\]]+(\.[^\s"(),.:;<>@[\\\]]+)*)|(".+"))@((\[(?:\d{1,3}\.){3}\d{1,3}])|(([\dA-Za-z-]+\.)+[A-Za-z]{2,}))$/;
```
Comment (Nizar-Rahme) [inline frontend/shared/src/constants.ts:16]: The e-mail regex on [this list from OWASP](https://owasp.org/www-community/OWASP_Validation_Regex_Repository) might be useful if you don't want to disable the safety rule.

**Review by Nizar-Rahme — COMMENTED** [review]
I left a few comments on the files I went through. I will pick it up later today from where I left.

File: frontend/kesaseteli/employer/src/components/application/form/ActionButtons.sc.ts (hunk 16 above)
Comment (Nizar-Rahme) [reply frontend/kesaseteli/employer/src/components/application/form/ActionButtons.sc.ts:6]: And margins specifically, I like to leave up to the parent component to determine its value.

File: frontend/kesaseteli/employer/src/__tests__/utils/components/render-page.tsx (hunk 15 above)
Comment (nenonja) [reply frontend/kesaseteli/employer/src/__tests__/utils/components/render-page.tsx:10]: For some reason prettier does not fix these

File: frontend/kesaseteli/employer/src/components/application/form/ActionButtons.sc.ts (hunk 16 above)
Comment (nenonja) [reply frontend/kesaseteli/employer/src/components/application/form/ActionButtons.sc.ts:6]: I dont know... I would rather keep the styles in sc.ts files and components more of the business logic and the dom structure. It's kinda tought that if I want to concentrate on business logic I work with components and hook. And when I want to fiddle with styles, then I look into sc.ts files.

File: frontend/kesaseteli/employer/src/components/application/steps/step2/error-notification/EmploymentsErrorNotification.sc.ts (hunk 18 above)
Comment (nenonja) [reply frontend/kesaseteli/employer/src/components/application/steps/step2/error-notification/EmploymentsErrorNotification.sc.ts:5]: Perhaps there was a thought of using more descriptive name for the component than GridCell but I can remove this.

File: frontend/shared/src/constants.ts (hunk 20 above)
Comment (nenonja) [reply frontend/shared/src/constants.ts:16]: Funny enough it still complains about the security but also about unicorn/better-regex. I took the advice from unicorn and it turn out to this: 
```
  // eslint-disable-next-line security/detect-unsafe-regex
  /^[\w&*+-]+(?:\.[\w&*+-]+)*@(?:[\dA-Za-z-]+\.)+[A-Za-z]{2,7}$/
```
It's shorter and probably safer though if it's from owasp.

File: frontend/shared/src/constants.ts (hunk 19 above)
Comment (nenonja) [reply frontend/shared/src/constants.ts:20]: I started to use these 👍

File: frontend/kesaseteli/employer/src/components/application/steps/step2/error-notification/EmploymentsErrorNotification.sc.ts (hunk 18 above)
Comment (nenonja) [reply frontend/kesaseteli/employer/src/components/application/steps/step2/error-notification/EmploymentsErrorNotification.sc.ts]: Removed

File: frontend/benefit/applicant/src/components/applications/forms/application/step2/ApplicationFormStep2.tsx [hunk 21]
```diff
@@ -478,7 +478,6 @@ const ApplicationFormStep2: React.FC<DynamicFormStepComponentProps> = ({
 
             <$GridCell $colSpan={12}>
               <Heading
-                as="h3"
```
Comment (Nizar-Rahme) [inline frontend/benefit/applicant/src/components/applications/forms/application/step2/ApplicationFormStep2.tsx:481]: Why did you have to remove it?

File: frontend/kesaseteli/employer/src/components/application/form/SelectionGroup.tsx [hunk 22]
```diff
@@ -0,0 +1,87 @@
...
+};
+
+SelectionGroup.defaultProps = {
+  direction: 'horizontal',
+  validation: undefined,
+  showTitle: true,
+  onChange: noop,
+}
+
+export default SelectionGroup;
```
Comment (Nizar-Rahme) [inline frontend/kesaseteli/employer/src/components/application/form/SelectionGroup.tsx:87]: I like these components. We might reduce a lot of repetitive code in Helsinki-Benefit by implementing a similar pattern.

File: frontend/kesaseteli/employer/src/components/application/form/DateInput.tsx [hunk 23]
```diff
@@ -0,0 +1,116 @@
...
+          validate,
+          setValueAs: convertDateForBackend,
+        })}
+        key={id}
+        id={id}
+        data-testid={id}
+        name={id}
+        required={Boolean(validation.required)}
+        initialMonth={new Date()}
+        defaultValue={date}
```
Comment (Nizar-Rahme) [inline frontend/kesaseteli/employer/src/components/application/form/DateInput.tsx:104]: I noticed that this component has the `defaultValue` prop but not the `value`. Is that intentional?

File: frontend/kesaseteli/employer/src/components/application/form/ActionButtons.sc.ts (hunk 16 above)
Comment (Nizar-Rahme) [reply frontend/kesaseteli/employer/src/components/application/form/ActionButtons.sc.ts:6]: Fair enough!

File: frontend/benefit/applicant/src/components/applications/forms/application/step2/ApplicationFormStep2.tsx (hunk 21 above)
Comment (nenonja) [reply frontend/benefit/applicant/src/components/applications/forms/application/step2/ApplicationFormStep2.tsx:481]: Probably an accident. I try to put it back.

File: frontend/kesaseteli/employer/src/components/application/form/SelectionGroup.tsx (hunk 22 above)
Comment (nenonja) [reply frontend/kesaseteli/employer/src/components/application/form/SelectionGroup.tsx:87]: Yes that would make the code much more readable 👍

File: frontend/benefit/applicant/src/components/applications/forms/application/step2/ApplicationFormStep2.tsx (hunk 21 above)
Comment (nenonja) [reply frontend/benefit/applicant/src/components/applications/forms/application/step2/ApplicationFormStep2.tsx:481]: reverted

File: frontend/kesaseteli/employer/src/components/application/form/DateInput.tsx (hunk 23 above)
Comment (nenonja) [reply frontend/kesaseteli/employer/src/components/application/form/DateInput.tsx:104]: I did lot of things back and forth. I think it's an accident. I replaced this with `value={date}` and it seem to work in the same way

File: frontend/benefit/applicant/src/components/applications/forms/application/step2/ApplicationFormStep2.tsx (hunk 21 above)
Comment (nenonja) [reply frontend/benefit/applicant/src/components/applications/forms/application/step2/ApplicationFormStep2.tsx]: Hmm. After merge it started to nag about it:
```
@frontend/applicant: Type error: Type '{ as: string; size: "xs"; header: string | undefined; tooltip: string | undefined; }' is not assignable to type 'IntrinsicAttributes & HeadingProps & { children?: ReactNode; }'.
@frontend/applicant:   Property 'as' does not exist on type 'IntrinsicAttributes & HeadingProps & { children?: ReactNode; }'.
@frontend/applicant:   536 |             <$GridCell $colSpan={12}>
@frontend/applicant:   537 |               <Heading
@frontend/applicant: > 538 |                 as="h3"
@frontend/applicant:       |                 ^
@frontend/applicant:   539 |                 size="xs"
@frontend/applicant:   540 |                 header={t(`${translationsBase}.heading5EmploymentSub1`)}
@frontend/applicant:   541 |                 tooltip={t(

```
I will check what to do with it.

Comment (nenonja) [reply frontend/benefit/applicant/src/components/applications/forms/application/step2/ApplicationFormStep2.tsx]: The Heading component's `as` property eas simply dropped out, so I added it back. Now it works.

File: frontend/kesaseteli/employer/src/components/application/form/TextInput.tsx [hunk 24]
```diff
@@ -1,47 +1,92 @@
...
+        as={getComponentType(type)}
+        {...register(id, {...validation, setValueAs: setValueForBackend})}
+        $type={type}
+        key={id}
+        id={id}
+        data-testid={id}
+        name={id}
+        required={Boolean(validation.required)}
+        max={validation.maxLength ? String(validation.maxLength) : undefined}
+        defaultValue={getValue()}
```
Comment (Nizar-Rahme) [inline frontend/kesaseteli/employer/src/components/application/form/TextInput.tsx:85]: Is this also an uncontrolled input on purpose?

File: frontend/kesaseteli/employer/src/components/application/form/TextInput.tsx [hunk 25]
```diff
@@ -1,47 +1,92 @@
...
+    return error;
+  },[t,errorType, helperFormat]);
+
+  // TODO: This can be removed after backend supports invalid values in draft save
+  const setValueForBackend = React.useCallback((newValue: string) =>
+    // getError does not always update: https://github.com/react-hook-form/react-hook-form/issues/2893
+    // if value hasnt changed (getValue is same as new value), then error is present and invalid value is changed to undefined
+    // to prevent backend to fail
+    getError() && getValue() === newValue ? undefined : newValue,
+   [getError,getValue]);
```
Comment (Nizar-Rahme) [inline frontend/kesaseteli/employer/src/components/application/form/TextInput.tsx:70]: You may run this file through Prettier.

File: frontend/kesaseteli/employer/src/components/application/form/DateInput.tsx [hunk 26]
```diff
@@ -0,0 +1,116 @@
...
+          ...validation,
+          validate,
+          setValueAs: convertDateForBackend,
+        })}
+        key={id}
+        id={id}
+        data-testid={id}
+        name={id}
+        required={Boolean(validation.required)}
+        initialMonth={new Date()}
```
Comment (Nizar-Rahme) [inline frontend/kesaseteli/employer/src/components/application/form/DateInput.tsx:103]: Maybe for later, you may need to define the `min` and `max` props.

File: frontend/kesaseteli/employer/src/hooks/backend/useLogoutQuery.ts [hunk 27]
```diff
@@ -13,6 +14,7 @@ const useLogoutQuery = (): UseMutationResult<unknown, Error, void> => {
     {
       onSuccess: () => {
         void queryClient.removeQueries();
+        clearLocalStorage('application');
```
Comment (Nizar-Rahme) [inline frontend/kesaseteli/employer/src/hooks/backend/useLogoutQuery.ts:17]: I think the key `application` should be either a constant or strictly typed. Otherwise, any future adjustments will keep the data stored in localStorage.

File: frontend/kesaseteli/employer/src/hooks/employments/useWatchEmployeeDisplayName.ts [hunk 28]
```diff
@@ -0,0 +1,14 @@
...
+import { isEmpty } from 'shared/utils/string.utils';
+
+const useWatchEmployeeDisplayName = (index: number): string => {
+  const { t } = useTranslation();
+  const defaultHeading = `${t(`common:application.step2.employment`)} #${
+    index + 1
+  }`;
+  const { watch: watchEmployeeName } = useApplicationFormField<string>(`summer_vouchers.${index}.employee_name`);
+  const employeeName = watchEmployeeName();
+  return !employeeName || isEmpty(employeeName) ? defaultHeading : employeeName;
```
Comment (Nizar-Rahme) [inline frontend/kesaseteli/employer/src/hooks/employments/useWatchEmployeeDisplayName.ts:12]: `isEmpty` can handle that, can't it?
```suggestion
  return isEmpty(employeeName) ? defaultHeading : employeeName;
```

**Comment by Nizar-Rahme** [comment]
Getting there 🐢 
image

File: frontend/kesaseteli/employer/src/components/application/steps/step2/accordions/EmploymentAccordion.sc.ts [hunk 29]
```diff
@@ -0,0 +1,28 @@
...
+  margin: ${({ theme: { spacing } }) =>
+    `${spacing.s} ${spacing.l} ${spacing.s} ${spacing.l}`};
+`;
+
+export const $Accordion = styled(Accordion)`
+  border: 1px solid ${({ theme: { colors } }) => colors.black50};
+`;
+export const $AccordionFormSection = styled(FormSection)`
+  background: ${(props) => props.theme.colors.black5};
+  border: 0px;
```
Comment (Nizar-Rahme) [inline frontend/kesaseteli/employer/src/components/application/steps/step2/accordions/EmploymentAccordion.sc.ts:17]:
```suggestion
  border: 0;
```

File: frontend/shared/src/utils/string.utils.ts [hunk 30]
```diff
@@ -23,3 +23,6 @@ export const getBooleanValueFromString = (value?: string): boolean | null => {
   }
   return true;
 };
+
+export const isEmpty = (value?: string): boolean =>
+  typeof value === 'string' ? value?.trim().length === 0 : Boolean(!value);
```
[the remark this answers is not in the document]
Comment (nenonja) [reply frontend/shared/src/utils/string.utils.ts:28]: The idea was to check the type runtime, then you need this `typeof value === 'string'`. But actually this method is obsolete because in later prs things get more complicate and I just replaced this with lodash isEmpty function.

File: frontend/kesaseteli/employer/src/components/application/form/TextInput.tsx (hunk 24 above)
Comment (nenonja) [reply frontend/kesaseteli/employer/src/components/application/form/TextInput.tsx:85]: Yes, react-hook-form embraces uncontrolled inputs... This does not work if using value instead of defaultValue.

File: frontend/kesaseteli/employer/src/components/application/form/TextInput.tsx (hunk 25 above)
Comment (nenonja) [reply frontend/kesaseteli/employer/src/components/application/form/TextInput.tsx:70]: prettier should be run on pre-commit hook and most of the times it works but somehow it fails sometimes. Not sure if there is something wrong with pre-hook or is it only on my local machine.

File: frontend/kesaseteli/employer/src/components/application/form/DateInput.tsx (hunk 26 above)
Comment (nenonja) [reply frontend/kesaseteli/employer/src/components/application/form/DateInput.tsx:103]: Yes we have a task to do it later... now on mvp we dont even check if start and end input are mixed wrong way.

File: frontend/kesaseteli/employer/src/hooks/employments/useWatchEmployeeDisplayName.ts (hunk 28 above)
Comment (nenonja) [reply frontend/kesaseteli/employer/src/hooks/employments/useWatchEmployeeDisplayName.ts:12]: yes it does but typescript does not catch it so the first term is for typescript only :(
afaik in the future versions of typescript this will be fixed https://devblogs.microsoft.com/typescript/announcing-typescript-4-4/#cfa-aliased-conditions
Actually 4.4 is now released but we are using version 4.3.4 I could try to update and remove the first term.

File: frontend/kesaseteli/employer/src/components/application/steps/step2/accordions/EmploymentAccordion.sc.ts (hunk 29 above)
Comment (nenonja) [reply frontend/kesaseteli/employer/src/components/application/steps/step2/accordions/EmploymentAccordion.sc.ts:17]: fixed

File: frontend/kesaseteli/employer/src/hooks/backend/useLogoutQuery.ts (hunk 27 above)
Comment (nenonja) [reply frontend/kesaseteli/employer/src/hooks/backend/useLogoutQuery.ts:17]: I started to use enum for it. Another place it is in use is here:

Comment (nenonja) [reply frontend/kesaseteli/employer/src/hooks/backend/useLogoutQuery.ts:17]: It's kind of hard because sometimes keys are dynamically created, like 
`clearLocalStorage(`application-${completeApplication.id}`)` It's kinda same as with react-querys invalidateQueries:
https://react-query.tanstack.com/guides/query-invalidation

File: frontend/kesaseteli/employer/src/hooks/employments/useWatchEmployeeDisplayName.ts (hunk 28 above)
Comment (nenonja) [reply frontend/kesaseteli/employer/src/hooks/employments/useWatchEmployeeDisplayName.ts:12]: seem to work

File: frontend/shared/src/components/forms/heading/Heading.tsx [hunk 31]
```diff
@@ -1,27 +1,16 @@
...
-  loadingText?: string;
-  loadingFinishedText?: string;
-  tooltip?: string;
-};
+import { $Header, HeadingProps } from './Heading.sc';
 
 const Heading: React.FC<HeadingProps> = ({
   as,
-  size,
+  size = 'l',
```
Comment (Nizar-Rahme) [inline frontend/shared/src/components/forms/heading/Heading.tsx:8]: Changing the default size from `'m'` to `'l'` may affect other components, have you checked that?

File: frontend/shared/src/utils/date.utils.ts [hunk 32]
```diff
@@ -32,21 +43,40 @@ export const formatDate = (
...
+  date: string | Date | number | undefined,
+  toFormat = DATE_FORMATS.BACKEND_DATE
+): string => {
+  const parsedDate = typeof date === 'string' ? parseDate(date) : date;
+  return formatDate(parsedDate, toFormat);
+};
+
+export const convertToUIDateFormat = (
+  date: string | Date | number | undefined
+): string => convertDateFormat(date, DATE_FORMATS.UI_DATE);
```
Comment (Nizar-Rahme) [inline frontend/shared/src/utils/date.utils.ts:79]:
```suggestion
): string => convertDateFormat(date, DATE_FORMATS.UI_DATE);

```

File: frontend/shared/src/components/accordion/Accordion.module.sc.ts [hunk 33]
```diff
@@ -0,0 +1,48 @@
...
+  padding-bottom: var(--spacing-m);
+  display: flex;
+  justify-content: space-between;
+  align-items: center;
+`;
+
+export const $AccordionContent = styled.div`
+  font-size: var(--fontsize-body-m);
+  line-height: var(--lineheight-l);
+  color: var(-color-black-90);
```
Comment (Nizar-Rahme) [inline frontend/shared/src/components/accordion/Accordion.module.sc.ts:39]:
```suggestion
  color: var(--color-black-90);
```

File: frontend/shared/src/components/accordion/Accordion.module.sc.ts [hunk 34]
```diff
@@ -0,0 +1,48 @@
...
+
+export const $AccordionHeader = styled.div<AccordionProps>`
+  position: relative;
+  background-color: ${(props) => props.headerBackgroundColor};
+  color: var(--color-black-90);
+  font-size: var(--fontsize-heading-m);
+  font-weight: bold;
+  line-height: var(--lineheight-m);
+  padding-top: var(--spacing-m);
+  padding-bottom: var(--spacing-m);
```
Comment (Nizar-Rahme) [inline frontend/shared/src/components/accordion/Accordion.module.sc.ts:30]: Why don't you use the mapped values from `theme` in this component?

**Review by Nizar-Rahme — COMMENTED** [review]
Last set of comments on this PR

File: frontend/kesaseteli/employer/src/components/application/steps/step2/error-notification/EmploymentsErrorNotification.tsx [hunk 35]
```diff
@@ -0,0 +1,69 @@
...
+    return null;
+  }
+  const employmentsErrors = errors.summer_vouchers || [];
+  const employmentErrorEntries = Array.isArray(employmentsErrors)
+    ? employmentsErrors.map((employmentErrors, index) => ({
+        index,
+        errors: Object.entries(employmentErrors ?? {}).map(
+          ([field, error]) => ({
+            field: field as keyof Employment,
+            errorType: error?.type || 'required',
```
[the remark this answers is not in the document]
Comment (nenonja) [reply frontend/kesaseteli/employer/src/components/application/steps/step2/error-notification/EmploymentsErrorNotification.tsx:37]: It seems that it would never fallback to that value because error object should not be undefined, I think I added it for the typescript only.  Also the errorType needs some of the react-hook-form error type value:
`errorType: LiteralUnion<"required" | "min" | "max" | "maxLength" | "minLength" | "pattern" | "validate" | "valueAsNumber" | "valueAsDate" | "value" | "setValueAs" | ..., string>
`

File: frontend/kesaseteli/employer/src/components/application/steps/step2/error-notification/EmployeeErrorNotification.tsx [hunk 36]
```diff
@@ -0,0 +1,34 @@
...
+};
+
+const EmployeeErrorNotification: React.FC<Props> = ({
+  index,
+  errors,
+}: Props) => {
+  const { t } = useTranslation();
+  const employeeDisplayname = useWatchEmployeeDisplayName(index);
+  return (
+    <$GridCell key={index}>
```
[the remark this answers is not in the document]
Comment (nenonja) [reply frontend/kesaseteli/employer/src/components/application/steps/step2/error-notification/EmployeeErrorNotification.tsx:23]: I'll remove it. I think there were some bug and react-hook-form docs recommended using key outside the loop and I tried to use it. Finally bug were somewhere else and this was left.

File: frontend/shared/src/components/forms/section/FormSection.sc.ts [hunk 37]
```diff
@@ -1,9 +1,20 @@
...
 type SpacingKeys = keyof DefaultTheme['spacing'];
 type SpacingValue = DefaultTheme['spacing'][SpacingKeys];
 
-type GridProps = {
+export type FormSectionProps = {
+  children?: React.ReactNode;
+  action?: React.ReactNode;
+  withoutDivider?: boolean;
+  header?: string;
+  'padding-bottom'?: boolean;
```
[the remark this answers is not in the document]
Comment (nenonja) [reply frontend/shared/src/components/forms/section/FormSection.sc.ts:12]: ESLint: ["paddingBottom"] is better written in dot notation.(@typescript-eslint/dot-notation)

[the remark this answers is not in the document]
Comment (nenonja) [reply frontend/shared/src/components/forms/section/FormSection.sc.ts:12]: will change this

File: frontend/shared/src/components/forms/heading/Heading.tsx (hunk 31 above)
Comment (nenonja) [reply frontend/shared/src/components/forms/heading/Heading.tsx:8]: I tried to remove it and couldn't see difference in my app. So I think it safer to remove, so no accidental changs to benefit.

File: frontend/shared/src/utils/date.utils.ts (hunk 32 above)
Comment (nenonja) [reply frontend/shared/src/utils/date.utils.ts:79]: fixing

File: frontend/shared/src/components/accordion/Accordion.module.sc.ts (hunk 34 above)
Comment (nenonja) [reply frontend/shared/src/components/accordion/Accordion.module.sc.ts:30]: dunno... I will fix it.

File: frontend/shared/src/components/forms/heading/Heading.tsx [hunk 38]
```diff
@@ -21,7 +10,7 @@ const Heading: React.FC<HeadingProps> = ({
   loading,
   tooltip,
 }) => (
-  <$Header size={size || 'm'} as={as}>
+  <$Header size={size} as={as}>
```
Comment (Nizar-Rahme) [inline frontend/shared/src/components/forms/heading/Heading.tsx:13]: One last thing, if we are going to remove this default value, I suggest making the size a required prop.

**Review by Nizar-Rahme — APPROVED** [review]
Great job! LGTM 👍

File: frontend/shared/src/components/forms/heading/Heading.tsx (hunk 38 above)
Comment (nenonja) [reply frontend/shared/src/components/forms/heading/Heading.tsx:13]: Reverting this too, or put it like `size = 'm',`
